db/write_batch.cc

```diff
@@ -1709,32 +1709,44 @@ class MemTableInserter : public WriteBatch::Handler {
       if (cf_handle == nullptr) {
         cf_handle = db_->DefaultColumnFamily();
       }
-      db_->Get(read_options, cf_handle, key, &get_value);
-      Slice get_value_slice = Slice(get_value);
-
-      // 2) Apply this merge
-      auto merge_operator = moptions->merge_operator;
-      assert(merge_operator);
-
-      std::string new_value;
-
-      Status merge_status = MergeHelper::TimedFullMerge(
-          merge_operator, key, &get_value_slice, {value}, &new_value,
-          moptions->info_log, moptions->statistics, Env::Default());
-
-      if (!merge_status.ok()) {
-        // Failed to merge!
+      Status get_status = db_->Get(read_options, cf_handle, key, &get_value);
+      if (!get_status.ok() && !get_status.IsNotFound()) {
+        // Failed to get!
         // Store the delta in memtable
         perform_merge = false;
       } else {
-        // 3) Add value to memtable
-        assert(!concurrent_memtable_writes_);
-        bool mem_res = mem->Add(sequence_, kTypeValue, key, new_value);
-        if (UNLIKELY(!mem_res)) {
-          assert(seq_per_batch_);
-          ret_status = Status::TryAgain("key+seq exists");
-          const bool BATCH_BOUNDRY = true;
-          MaybeAdvanceSeq(BATCH_BOUNDRY);
+        Slice get_value_slice = Slice(get_value);
+        Slice* merge_data;
+        if (get_status.ok()) {
+          merge_data = &get_value_slice;
+        } else {  // Key not present in db. (s.IsNotFound())
+          merge_data = nullptr;
+        }
+
+        // 2) Apply this merge
+        auto merge_operator = moptions->merge_operator;
+        assert(merge_operator);
+
+        std::string new_value;
+
+        Status merge_status = MergeHelper::TimedFullMerge(
+            merge_operator, key, merge_data, {value}, &new_value,
```

Comment:
I'm not sure whether I got it correctly. Now `merge_data` can be null but previously this argument cannot be null. And this argument can be passed all the way into users' merge operator. Is it an API change to merge operator?

Reply:
The inline comment here says nullptr should be passed if the value does not exists or is delete tombstone (https://github.com/facebook/rocksdb/blob/master/include/rocksdb/merge_operator.h#L57). This is consistent with other call sites like this one: https://github.com/facebook/rocksdb/blob/a19de78da5b8e0ab5ddbcada60e473b84d4dc72f/db/db_iter.cc#L519

Reply:
You are right about the interface.

Another question, if there is no previous value, can we directly turning the value into Put() without invoking merge operator?

Reply:
It seems to me we cannot. Having merge operand being a partial value and have the merge operator turn it into full value seems to be a logical use of merge operator. In that case turning merge operand directly into value is not correct.

Reply:
OK.

I have another question. In which case it returns not found? The previous `num_merges >= moptions->max_successive_merges` check means that there is at least one value in memtable. I guess it returns not found because of range deletes?

Since we are here, can you add a unit test to cover this code path?

Reply:
good point, that's why it didn't check not found case.. How about we simply reset `perform_merge` to false if we get not found here?

Reply:
Yes, perform_merge = true for NotFound case sounds reasonable.